### DISCUSSION TITLE
batch mode and mocking support

### DISCUSSION THREAD

**Discussion by nkiesel** [komapper/komapper#612] [Ideas]
[Full disclosure: I'm the author of https://github.com/MetricStream/SQLBuilder, a much more primitive JDBC support framework].

Komapper is coming along nicely. One production code aspect that I do not yet see covered is "batch mode" support. Do you have any plans for that?

The other missing aspect is mocking support for tests. This was a big issue for our usage because our legacy code has queries spread throughout application code.  Standard mockk / (Power)Mockito mocking works, but has many downsides (e.g. it's very hard to mock results from specific queries). We thus built a full mocking implementation for our framework.  Have you thought on how to support mocking and unit testing for Komapper?

**Comment by nakamura-to** [comment]
Thanks for your interest in Komapper.

> One production code aspect that I do not yet see covered is "batch mode" support. Do you have any plans for that?

We support [batch insert](https://www.komapper.org/docs/reference/query/querydsl/insert/#batch), [batch update](https://www.komapper.org/docs/reference/query/querydsl/update/#batch), and [batch delete](https://www.komapper.org/docs/reference/query/querydsl/delete/#batch).

> Have you thought on how to support mocking and unit testing for Komapper?

We believe that using a repository pattern is a good idea:

```kotlin
interface EmployeeRepository {
    fun findByName(name: String): List<Employee>
}

class EmployeeRepositoryImpl(private val db: JdbcDatabase): EmployeeRepository {
    override fun findByName(name: String): List<Employee> {
        val e = Meta.employee
        return db.runQuery { 
            QueryDsl.from(e).where { e.name eq name}
        }
    }
}
```

In the above example, mocking the `EmployeeRepository` interface is easy.

**Reply by nkiesel** [reply]
Your batch mode support seems to require that all the data is pre-computed. We often have cases where we create batches using streams/sequences of data (e.g. comping form another DB connection).  I'm currently exploring to use per-column generators (with implementations providing constants, values from collections, values from streams/sequences, values from  lambdas, ....) for our SQLBuilder project.

Regarding testing: mocking the repo looks good, but in my experience the real challenge is to provide mocked data in a structured manner. One common issue with the usual mockk/mockito mocking is that the mocks provide an unstructured flow of data (`return(3).andThen(4)...`).  This is challenging if the tested code runs multiple different queries, and very easily leads to "off-by-one" errors (e.g. someone added an extra query deep in the tested code). My approach is to thus provide explicit records of mocked data and "tag" them with the name of the method in which the data is supposed to be consumed.  My mocking driver then enforces this constraint.
